nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +530,369 @@ impl DataStore {
 
         Ok(())
     }
+
+    /// Attempts to lock an instance's record to apply state updates in an
+    /// instance-update saga, returning the state of the instance when the lock
+    /// was acquired.
+    ///
+    /// # Notes
+    ///
+    /// This method MUST only be called from the context of a saga! The
+    /// calling saga must ensure that the reverse action for the action that
+    /// acquires the lock must call [`DataStore::instance_updater_unlock`] to
+    /// ensure that the lock is always released if the saga unwinds.
+    ///
+    /// # Arguments
+    ///
+    /// - `authz_instance`: the instance to attempt to lock to lock
+    /// - `saga_lock_id`: the UUID of the saga that's attempting to lock this
+    ///   instance.
+    ///
+    /// # Returns
+    ///
+    /// - [`Ok`]`(`[Instance`]`)` if the lock was acquired.
+    /// - [`Err`]`([`Error::InternalError`])` if the instance was locked by
+    ///   another saga.
+    /// - [`Err`]`(`[`Error::ObjectNotFound`]`)` if no instance with the
+    ///   provided ID exists (or the instance record has been deleted).
+    /// - [`Err`]`(other)` if the database query failed for another reason.
+    pub async fn instance_updater_lock(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        saga_lock_id: &Uuid,
```

Comment:
Hmm, yeah...I also thought about having it return `Result<Option<Instance>,
Error>` to distinguish between "you can't get ye lock" and "something unrelated
went wrong", but I felt like the "lock not acquired" case should _probably_ be
a `Result` so that you can't easily ignore it...